sign/dilithium/mode3/internal/sample.go

```diff
@@ -2,9 +2,190 @@ package internal
+
+	for !done {
+		// Applies KeccaK-f[1600] to state to get the next 17 uint64s of each
+		// of the four SHAKE256 streams.
+		perm.Permute()
+
+		done = true
+
+		for j := 0; j < 4; j++ {
+			if idx[j] == 256 {
+				continue
+			}
+			done = false
+
+			for i := 0; i < 17; i++ {
+				var t [4]uint32
+				tCount := 3
+
+				// Get the next three or four 20 bit numbers.
+				qw := state[i*4+j]
+				qwl := (qw << shift[j]) | uint64(carry[j])
+				t[0] = uint32(qwl & 0xfffff)
+				t[1] = uint32((qwl >> 20) & 0xfffff)
+				t[2] = uint32((qwl >> 40) & 0xfffff)
+
+				if shift[j] == 16 {
+					t[3] = uint32(qw >> 44)
+					shift[j] = 0
+					carry[j] = 0
+					tCount = 4
+				} else {
+					shift[j] += 4
+					carry[j] = uint32(qw >> (64 - shift[j]))
+				}
+
+				// Check if they're coefficients.
+				for k := 0; k < tCount; k++ {
+					if t[k] <= 2*common.Gamma1-2 {
+						ps[j][idx[j]] = common.Q + common.Gamma1 - 1 - t[k]
+						idx[j]++
+						if idx[j] == 256 {
+							break
+						}
+					}
+				}
+
+				if idx[j] == 256 {
+					break
+				}
+			}
+		}
+	}
+}
+
+// For each i, sample ps[i] uniformly from the given seed and nonces[i].
+// ps[i] may be nil and is ignored in that case.
+//
+// Can only be called when DeriveX4Available is true.
+func PolyDeriveUniformX4(ps [4]*common.Poly, seed *[32]byte, nonces [4]uint16) {
+	var perm f1600x4.State
+	state := perm.Initialize()
+
+	// Absorb the seed in the four states
+	for i := 0; i < 4; i++ {
+		v := binary.LittleEndian.Uint64(seed[8*i : 8*(i+1)])
+		for j := 0; j < 4; j++ {
+			state[i*4+j] = v
+		}
+	}
+
+	// Absorb the nonces, the SHAKE128 domain separator (0b1111), the
+	// start of the padding (0b...001) and the end of the padding 0b100...
+	// Recall that the rate of SHAKE128 is 168 --- i.e. 21 uint64s.
+	for j := 0; j < 4; j++ {
+		state[4*4+j] = uint64(nonces[j]) | (0x1f << 16)
+		state[20*4+j] = 0x80 << 56
+	}
+
+	var idx [4]int // indices into ps
+	for j := 0; j < 4; j++ {
+		if ps[j] == nil {
+			idx[j] = 256 // mark nil polynomial as completed
+		}
+	}
+
+	for {
+		// Applies KeccaK-f[1600] to state to get the next 21 uint64s of each
+		// of the four SHAKE128 streams.
+		perm.Permute()
+
+		done := true
+
+		for j := 0; j < 4; j++ {
+			if idx[j] == 256 {
+				continue
+			}
+			done = false
+			for i := 0; i < 7; i++ {
+				var t [8]uint32
+				t[0] = uint32(state[i*3*4+j] & 0x7fffff)
+				t[1] = uint32((state[i*3*4+j] >> 24) & 0x7fffff)
+				t[2] = uint32((state[i*3*4+j] >> 48) |
+					((state[(i*3+1)*4+j] & 0x7f) << 16))
+				t[3] = uint32((state[(i*3+1)*4+j] >> 8) & 0x7fffff)
+				t[4] = uint32((state[(i*3+1)*4+j] >> 32) & 0x7fffff)
+				t[5] = uint32((state[(i*3+1)*4+j] >> 56) |
+					((state[(i*3+2)*4+j] & 0x7fff) << 8))
+				t[6] = uint32((state[(i*3+2)*4+j] >> 16) & 0x7fffff)
+				t[7] = uint32((state[(i*3+2)*4+j] >> 40) & 0x7fffff)
+
+				for k := 0; k < 8; k++ {
+					if t[k] < common.Q {
+						ps[j][idx[j]] = t[k]
+						idx[j]++
+						if idx[j] == 256 {
+							break
```

Comment:
Take a look on "break with labels", it might be possible to write clearer the code.